vitessce/data_utils/anndata.py

```diff
@@ -0,0 +1,163 @@
+        cast_sum = np.sum(cast_arr)
+        if np.abs(orig_sum - cast_sum) < 1e-2:
+            arr = cast_arr
+
+    # Try casting signed int to unsigned int.
+    if arr.dtype.kind == 'i':
+        # If signed int dtype but no values are less than zero,
+        # convert to unsigned int dtype.
+        if arr.min() >= 0:
+            arr = arr.astype(f'<u{arr.dtype.itemsize}')
+
+    # Try casting to a smaller itemsize.
+    if arr.dtype.kind == 'u' or arr.dtype.kind == 'i' or arr.dtype.kind == 'f':
+        next_itemsizes = [4] if arr.dtype.kind == 'f' else [4, 2, 1]
+        for next_itemsize in next_itemsizes:
+            if arr.dtype.itemsize > next_itemsize:
+                next_dtype = np.dtype(f'<{arr.dtype.kind}{next_itemsize}')
+                next_dtype_info = np.iinfo(next_dtype) if arr.dtype.kind == 'u' or arr.dtype.kind == 'i' else np.finfo(next_dtype)
+                if next_dtype_info.min <= orig_min and next_dtype_info.max >= orig_max:
+                    arr = arr.astype(next_dtype)
+                elif arr.dtype.itemsize == 8 and (arr.dtype.kind == 'u' or arr.dtype.kind == 'i'):
+                    print(f"WARNING: Not casting array with dtype {arr.dtype.name}, but Zarr.js suggests avoiding int64 and uint64")
+
+    # Check for float16 usage.
+    if arr.dtype.kind == 'f' and arr.dtype.itemsize == 2:
+        # Zarr.js does not have a Float16Array type
+        arr = arr.astype('<f4')
+
+    return arr
+
+
+def optimize_arr(arr):
+    return to_dense(cast_arr(to_memory(arr)))
+
+
+# Given an anndata object, optimize for usage with Vitessce
+def optimize_adata(adata, obs_cols=None, obsm_keys=None, var_cols=None, varm_keys=None, layer_keys=None, ignore_X=False):
+    if not ignore_X:
+        if adata.X is not None:
+            new_X = optimize_arr(adata.X)
+        else:
+            new_X = adata.X
+    else:
+        new_X = None
+    # Only keep the subset of columns required.
+    if obs_cols is not None:
+        new_obs = adata.obs[obs_cols]
+    else:
+        new_obs = adata.obs
+    if var_cols is not None:
+        new_var = adata.var[var_cols]
+    else:
+        new_var = adata.var
+    # Only keep the subset of obsm and varm items required.
+    if obsm_keys is None:
+        obsm_keys = adata.obsm.keys()
+    new_obsm = {
+        obsm_key: optimize_arr(adata.obsm[obsm_key])
+        for obsm_key
+        in obsm_keys
+    }
+    if varm_keys is None:
+        varm_keys = adata.varm.keys()
+    new_varm = {
+        varm_key: optimize_arr(adata.varm[varm_key])
+        for varm_key
+        in varm_keys
+    }
+    if layer_keys is None:
+        layer_keys = adata.layers.keys()
+    new_layers = {
+        layer_key: optimize_arr(adata.layers[layer_key])
+        for layer_key
+        in layer_keys
+    }
+    adata = AnnData(X=new_X, obs=new_obs, var=new_var, obsm=new_obsm, varm=new_varm, layers=new_layers)
+    return adata
+
+
+def to_memory(arr):
+    if type(arr).__name__ == "SparseDataset":
```

Comment:
Could you point me to where you found this?  I am very curious since I have run in to some funny business with this sometimes.

Reply:
This comes from using a backed AnnData object (e.g. `read_h5ad(path, backed="r+")`). It is defined at https://github.com/scverse/anndata/blob/286bc7f207863964cb861f17c96ab24fe0cf72ac/anndata/_core/sparse_dataset.py#L230